In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv("/content/WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [3]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


# 01 - Dataset Understanding

## Dataset: Telco Customer Churn

This notebook focuses on understanding the structure and characteristics of the Telco Customer Churn dataset before performing data preprocessing or machine learning.

### Objectives

- Understand what a dataset is
- Identify features and target variables
- Identify independent and dependent variables
- Identify numerical and categorical features
- Identify discrete and continuous features
- Identify identifier columns
- Understand metadata
- Understand dataset dimensions

## 1. What is a Dataset?

A dataset is a structured collection of related data organized into rows and columns.

- Each row represents one observation or record.
- Each column represents a feature or attribute of that observation.

### Simple Example

A student dataset may contain:

| Student_ID | Age | Study_Hours | Result |
|------------|-----|-------------|--------|
| 101 | 20 | 5 | Pass |
| 102 | 21 | 3 | Fail |

Here, each row represents one student and each column represents information about the student.

### Real-World Example

A telecom company can maintain customer information such as tenure, contract type, internet service, monthly charges, and whether the customer churned.

### Business Example

A telecom company can analyze customer data to identify customers who are likely to leave the company.

### AI/ML Use Case

A dataset can be used to train a machine learning model to predict customer churn.

### Why is it important?

Understanding the dataset helps us identify the available information, data types, target variable, and potential data-quality issues before building an ML pipeline.

In [5]:
print("Shape:", df.shape)
print("\nColumns:")
for i, col in enumerate(df.columns, 1):
    print(i, ":", col)

print("\nData Types:")
print(df.dtypes)

Shape: (7043, 21)

Columns:
1 : customerID
2 : gender
3 : SeniorCitizen
4 : Partner
5 : Dependents
6 : tenure
7 : PhoneService
8 : MultipleLines
9 : InternetService
10 : OnlineSecurity
11 : OnlineBackup
12 : DeviceProtection
13 : TechSupport
14 : StreamingTV
15 : StreamingMovies
16 : Contract
17 : PaperlessBilling
18 : PaymentMethod
19 : MonthlyCharges
20 : TotalCharges
21 : Churn

Data Types:
customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object


## 2. What is a Feature?

A feature is an individual variable or attribute that describes an observation in a dataset.

In this Telco Customer Churn dataset, the customer information is represented through different features such as:

- gender
- SeniorCitizen
- Partner
- Dependents
- tenure
- PhoneService
- MultipleLines
- InternetService
- Contract
- MonthlyCharges
- TotalCharges

The dataset contains 21 columns in total. These include customer attributes, an identifier column, and the target variable.

### Simple Example

In a student dataset, `Age`, `Study_Hours`, and `Attendance` can be features.

### Real-World Example

In a telecom dataset, `tenure`, `Contract`, `MonthlyCharges`, and `InternetService` describe customer characteristics.

### Business Example

A telecom company can analyze these customer features to understand customer behavior and identify factors related to customer churn.

### AI/ML Use Case

Features can be used as input variables for a machine learning model that predicts whether a customer will churn.

### Why is it important?

Understanding features helps us identify what information is available in the dataset and which variables can be used for analysis and machine learning.

In [6]:
# Display all columns/features in the dataset

for i, column in enumerate(df.columns, 1):
    print(i, ":", column)

1 : customerID
2 : gender
3 : SeniorCitizen
4 : Partner
5 : Dependents
6 : tenure
7 : PhoneService
8 : MultipleLines
9 : InternetService
10 : OnlineSecurity
11 : OnlineBackup
12 : DeviceProtection
13 : TechSupport
14 : StreamingTV
15 : StreamingMovies
16 : Contract
17 : PaperlessBilling
18 : PaymentMethod
19 : MonthlyCharges
20 : TotalCharges
21 : Churn


## 3. What is a Target?

A target variable is the variable that we want to predict or explain using the other variables in the dataset.

For this dataset:

- Target variable: `Churn`
- Possible values: `Yes` and `No`

The target indicates whether a customer has left the telecom service.

In a machine learning problem, the target is usually represented as `y`, while the input features are represented as `X`.

In [7]:
# Display the target variable

df["Churn"].value_counts()

,count
Churn,
No,5174
Yes,1869


## 4. Independent Variables

Independent variables are the input variables used to explain or predict the target variable.

In this dataset, the independent variables include customer information such as:

- gender
- SeniorCitizen
- Partner
- Dependents
- tenure
- PhoneService
- MultipleLines
- InternetService
- OnlineSecurity
- OnlineBackup
- DeviceProtection
- TechSupport
- StreamingTV
- StreamingMovies
- Contract
- PaperlessBilling
- PaymentMethod
- MonthlyCharges
- TotalCharges

`customerID` is an identifier and should not normally be treated as a predictive feature.

The independent variables are generally represented by `X` in machine learning.

In [8]:
# Create independent variables by excluding the identifier and target

X = df.drop(columns=["customerID", "Churn"])

X.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
0,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85
1,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5
2,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15
3,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75
4,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65


## 5. Dependent Variable

A dependent variable is the variable whose value depends on or is explained by other variables.

In this dataset, `Churn` is the dependent variable because we want to understand or predict whether a customer will leave the telecom service.

The dependent variable is generally represented by `y` in machine learning.

In [9]:
# Create the dependent/target variable

y = df["Churn"]

y.head()

,Churn
0,No
1,No
2,Yes
3,No
4,Yes


## 6. Numerical Features

Numerical features are variables whose values are represented using numbers and can be used for mathematical calculations.

In this dataset, the columns stored with numerical data types are:

- `SeniorCitizen`
- `tenure`
- `MonthlyCharges`

`TotalCharges` represents a numeric quantity, but in our dataset it is stored as `object` because it contains blank values. Therefore, it is a data-quality issue that we will handle during data preprocessing.

### Simple Example

Examples of numerical features include age, salary, height, and number of years.

### Real-World Example

In telecom data, `tenure` represents the number of months a customer has stayed with the company, while `MonthlyCharges` represents the customer's monthly charge.

### Business Example

A telecom company can analyze monthly charges and customer tenure to understand customer behavior.

### AI/ML Use Case

Numerical features can be used as input variables for machine learning models to predict customer churn.

### Why is it important?

Identifying numerical features helps us select appropriate statistical analysis, visualizations, and preprocessing techniques.

In [10]:

# Identify columns with numerical data types

numerical_columns = df.select_dtypes(include=np.number).columns.tolist()

numerical_columns

['SeniorCitizen', 'tenure', 'MonthlyCharges']

### Important Observation

`TotalCharges` is not included in the above result because Pandas has detected it as an `object` column.

Although `TotalCharges` contains numerical information, its current data type is `object`. This indicates that the column requires further data-quality investigation before it can be treated as a numerical feature.

In [11]:
# Check the data type of TotalCharges

df["TotalCharges"].dtype

dtype('O')

## 7. Categorical Features

Categorical features contain values that represent groups or categories rather than continuous numerical measurements.

In this dataset, the following columns are categorical:

- `gender`
- `Partner`
- `Dependents`
- `PhoneService`
- `MultipleLines`
- `InternetService`
- `OnlineSecurity`
- `OnlineBackup`
- `DeviceProtection`
- `TechSupport`
- `StreamingTV`
- `StreamingMovies`
- `Contract`
- `PaperlessBilling`
- `PaymentMethod`
- `Churn`

`customerID` is also stored as `object`, but it is an identifier column rather than a meaningful categorical feature.

### Simple Example

Examples of categorical variables include gender, city, department, and payment method.

### Real-World Example

In telecom data, `Contract` contains categories such as month-to-month, one year, and two year.

### Business Example

A telecom company can compare churn rates across different contract types.

### AI/ML Use Case

Categorical features can be encoded into numerical representations before being given to many machine learning algorithms.

### Why is it important?

Identifying categorical features helps us choose appropriate analysis, visualization, and encoding techniques.

In [12]:
# Identify object columns

categorical_columns = df.select_dtypes(include="object").columns.tolist()

categorical_columns

['customerID',
 'gender',
 'Partner',
 'Dependents',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'TotalCharges',
 'Churn']

## 8. Discrete Features

Discrete features are numerical variables that contain separate, countable values.

In the Telco Customer Churn dataset:

- `SeniorCitizen` is a discrete feature because it contains only `0` and `1`.
- `tenure` is a discrete feature because it represents the number of months a customer has stayed with the company.

### Simple Example

Number of students in a classroom is a discrete variable because it can be counted.

### Real-World Example

Customer tenure measured in complete months is discrete because the values are represented as separate month counts.

### Business Example

A telecom company can analyze how churn changes as the customer's tenure increases.

### AI/ML Use Case

Discrete numerical features can be used as input variables in machine learning models.

### Why is it important?

Identifying discrete features helps us understand the nature of numerical variables and select suitable statistical analysis and visualizations.

In [13]:
# Check unique values of discrete features

print("SeniorCitizen:")
print(df["SeniorCitizen"].unique())

print("\nTenure:")
print(df["tenure"].unique()[:20])

SeniorCitizen:
[0 1]

Tenure:
[ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30]


## 9. Continuous Features

Continuous features are numerical variables that can take a range of values, including decimal values.

In the Telco Customer Churn dataset:

- `MonthlyCharges` is a continuous feature because it can contain decimal values.
- `TotalCharges` represents a continuous numerical quantity, although it is currently stored as `object` in the dataset.

### Simple Example

Height and weight are examples of continuous variables because they can take decimal values.

### Real-World Example

A customer's monthly telecom charge can be a decimal value such as 70.35.

### Business Example

A telecom company can analyze monthly charges to understand customer spending patterns.

### AI/ML Use Case

Continuous features can be used as numerical inputs for machine learning models.

### Why is it important?

Identifying continuous features helps determine appropriate statistical methods, visualizations, and preprocessing techniques.

In [14]:
# Check values of the continuous feature

df["MonthlyCharges"].head(10)

,MonthlyCharges
0,29.85
1,56.95
2,53.85
3,42.30
4,70.70
5,99.65
6,89.10
7,29.75
8,104.80
9,56.15


In [15]:
# Check the data type of TotalCharges

df["TotalCharges"].dtype

dtype('O')

## 10. Identifier Columns

An identifier column is a column used to uniquely identify each record or observation in a dataset.

In the Telco Customer Churn dataset:

- `customerID` is the identifier column.
- Each customer has a unique `customerID`.
- It is used to identify customers rather than describe their characteristics.

### Simple Example

In a student dataset, `Student_ID` can be used to uniquely identify each student.

### Real-World Example

A telecom company assigns a unique customer ID to each customer.

### Business Example

The company can use `customerID` to locate a particular customer's record.

### AI/ML Use Case

An identifier generally should not be used as a predictive feature because it does not contain meaningful information about customer behavior.

### Why is it important?

Identifying ID columns prevents meaningless or unique values from being incorrectly used as input features in a machine learning model.

In [16]:
# Check the identifier column

df["customerID"].head()

,customerID
0,7590-VHVEG
1,5575-GNVDE
2,3668-QPYBK
3,7795-CFOCW
4,9237-HQITU


In [17]:
# Check whether customerID values are unique

df["customerID"].nunique()

7043

### Observation

The dataset contains 7,043 records and the `customerID` column contains 7,043 unique values.

Therefore, `customerID` acts as a unique identifier for each customer.

## 11. Target Variable

The target variable is the variable that we want to predict or analyze using the other variables in the dataset.

In the Telco Customer Churn dataset:

- Target variable: `Churn`
- Possible values: `Yes` and `No`

`Churn` indicates whether a customer has left the telecom service.

For a machine learning problem:

- `X` represents the independent/input variables.
- `y` represents the dependent/target variable.

### Simple Example

In a house-price prediction dataset, `Price` can be the target variable.

### Real-World Example

In this dataset, we want to predict whether a customer will churn.

### Business Example

A telecom company can use churn predictions to identify customers who are likely to leave and take retention actions.

### AI/ML Use Case

`Churn` can be used as the target variable for a binary classification model.

### Why is it important?

Identifying the target variable defines what the machine learning model is expected to predict.

In [19]:
# Check the target variable

print("Target variable:", "Churn")
print("\nUnique values:")
print(df["Churn"].unique())

print("\nValue counts:")
print(df["Churn"].value_counts())

Target variable: Churn

Unique values:
['No' 'Yes']

Value counts:
Churn
No     5174
Yes    1869
Name: count, dtype: int64


## 12. Metadata

Metadata is information that describes the structure, characteristics, and meaning of a dataset.

For this dataset, metadata includes:

- Dataset name
- Number of rows
- Number of columns
- Column names
- Data types
- Meaning of columns
- Target variable
- Identifier column
- Data source

In [20]:
# Basic metadata of the dataset

print("Dataset Name: Telco Customer Churn")
print("Number of Rows:", df.shape[0])
print("Number of Columns:", df.shape[1])
print("Target Variable: Churn")
print("Identifier Column: customerID")

Dataset Name: Telco Customer Churn
Number of Rows: 7043
Number of Columns: 21
Target Variable: Churn
Identifier Column: customerID


## 13. Dataset Dimensions

Dataset dimensions describe the number of rows and columns present in a dataset.

- Rows represent individual observations or records.
- Columns represent variables or attributes.

For this dataset:

- Rows: 7,043
- Columns: 21

Therefore, the dataset contains 7,043 customer records and 21 columns.

In [21]:
# Check dataset dimensions

df.shape

(7043, 21)

### Dataset Structure Summary

The Telco Customer Churn dataset contains 7,043 customer records and 21 columns.

- `customerID` is the identifier column.
- `Churn` is the target variable.
- `SeniorCitizen`, `tenure`, and `MonthlyCharges` are detected as numerical columns.
- Several columns are categorical.
- `TotalCharges` represents numerical information but is currently stored as an object data type.